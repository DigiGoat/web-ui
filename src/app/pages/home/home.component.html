<div class="w-100 text-center">
  <img ngSrc="./assets/icons/android-chrome-192x192.png" alt="Farm Logo" width="300" height="300" class="text-center" priority>
</div>
<h1 class="text-center" test-id="title">{{config.homeTitle}}</h1>
<h5 class="text-center" test-id="owner">{{config.owner}}</h5>
<h6 class="text-center" test-id="email"><a [href]="'mailto:' + config.email">{{config.email}}</a></h6>
<div class="container">
  <p class="text-center" [innerHTML]="description" test-id="description"></p>
</div>
