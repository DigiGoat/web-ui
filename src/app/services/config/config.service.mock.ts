import type { ConfigService } from './config.service';

export const ConfigServiceMock = {
  homeTitle: 'TEST_TITLE',
  owner: 'TEST_OWNER',
  email: 'TEST_EMAIL',
  homeDescription: 'TEST_DESCRIPTION',
  menubarTitle: 'TEST_MENUBAR_TITLE',
  tabTitle: 'TEST_TAB_TITLE',
  link: 'TEST_LINK'
} as ConfigService;
export const EmptyConfigServiceMock = {
  homeTitle: '',
  owner: '',
  email: '',
  homeDescription: '',
  menubarTitle: '',
  tabTitle: '',
  link: ''
} as ConfigService;
