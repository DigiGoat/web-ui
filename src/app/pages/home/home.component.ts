import { Component, ElementRef, type OnInit } from '@angular/core';

import { Meta } from '@angular/platform-browser';
import type { Observable } from 'rxjs';
import { ConfigService } from '../../services/config/config.service';
import { GoatService } from '../../services/goat/goat.service';
import { PlatformService } from '../../services/platform/platform.service';


@Component({
  selector: 'app-home',
  templateUrl: './home.component.html',
  styleUrls: ['./home.component.scss'],
  standalone: false
})
export class HomeComponent implements OnInit {
  public description = '';
  constructor(public config: ConfigService, private meta: Meta, private goatService: GoatService, private platformService: PlatformService, private el: ElementRef<HTMLElement>) { }
  ngOnInit(): void {
    this.description = this.config.homeDescription;
    this.setDescription();
  }
  htmlToPlainText(html: string): string {
    const doc = this.el.nativeElement.ownerDocument.createElement('div');
    doc.innerHTML = html;
    return doc.textContent || '';
  }
  setDescription(): void | Observable<void> {
    if (this.description) {
      this.meta.addTags([{ property: 'og:description', content: this.htmlToPlainText(this.description) }, { name: 'description', content: this.htmlToPlainText(this.description) }]);
    } else {
      let description = '';
      if (this.config.title && this.config.owner) {
        description += `${this.config.title} is owned and operated by ${this.config.owner}. `;
      }
      this.goatService.does.subscribe(does => {
        this.goatService.bucks.subscribe(bucks => {

          if (this.config.title) {
            if (this.config.owner) {
              description += `The herd is currently home to ${does.length} doe${does.length === 1 ? '' : 's'} and ${bucks.length} buck${bucks.length === 1 ? '' : 's'}. `;
            } else {
              description += `${this.config.title} is currently home to ${does.length} doe${does.length === 1 ? '' : 's'} and ${bucks.length} buck${bucks.length === 1 ? '' : 's'}. `;
            }
          } else {
            description += `The herd is currently home to ${does.length} doe${does.length === 1 ? '' : 's'} and ${bucks.length} buck${bucks.length === 1 ? '' : 's'}. `;
          }
          this.meta.addTags([{ property: 'og:description', content: description }, { name: 'description', content: description }]);
        });
      });
    }
  }
}
