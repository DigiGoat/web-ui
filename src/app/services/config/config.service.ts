import { Injectable } from '@angular/core';

import config from '../../../assets/resources/config.json';


@Injectable({
  providedIn: 'root'
})
export class ConfigService {
  private config: Record<string, string | boolean | Record<string, string | Record<string, string>>> = config;
  get title(): string {
    if (this.config['title']) {
      return this.config['title'] as string;
    } else if (this.config['homeTitle']) {
      return this.config['homeTitle'] as string;
    } else if (this.config['menubarTitle']) {
      return this.config['menubarTitle'] as string;
    }
    return '';
  }
  get shortTitle(): string {
    if (this.config['shortTitle']) {
      return this.config['shortTitle'] as string;
    } else if (this.config['tabTitle']) {
      return this.config['tabTitle'] as string;
    }
    return '';
  }
  get owner(): string {
    if (this.config['owner']) {
      return this.config['owner'] as string;
    }
    return '';
  }
  get email(): string {
    if (this.config['email']) {
      return this.config['email'] as string;
    }
    return '';
  }
  get homeDescription(): string {
    if (this.config['homeDescription']) {
      return this.config['homeDescription'] as string;
    }
    return '';
  }
  get link(): string {
    if (this.config['link']) {
      return `${this.config['link']}${(this.config['link'] as string).endsWith('/') ? '' : '/'}`;
    }
    return '';
  }
  get analytics(): Analytics {
    if (this.config['analytics']) {
      return this.config['analytics'] as Analytics;
    }
    return {};
  }
  get colors(): ColorScheme {
    if (this.config['colors']) {
      return this.config['colors'] as ColorScheme;
    }
    return {};
  }
  get kiddingSchedule(): boolean {
    if (this.config['kiddingSchedule']) {
      return this.config['kiddingSchedule'] as boolean;
    }
    return false;
  }
  get kiddingScheduleDescription(): string {
    if (this.config['kiddingScheduleDescription']) {
      return this.config['kiddingScheduleDescription'] as string;
    }
    return '';
  }
  get socials(): Socials {
    if (this.config['socials']) {
      return this.config['socials'] as Socials;
    }
    return {};
  }
  get references(): boolean {
    if (this.config['references']) {
      return this.config['references'] as boolean;
    }
    return false;
  }
  get forSale(): boolean {
    if (this.config['forSale']) {
      return this.config['forSale'] as boolean;
    }
    return false;
  }
  get saleTerms(): string {
    if (this.config['saleTerms']) {
      return this.config['saleTerms'] as string;
    }
    return '';
  }
  get firebase(): Firebase | false {
    if (this.config['firebase'] && typeof this.config['firebase'] === 'object') {
      // Make sure alll firebase keys are present
      if (!this.config['firebase']['apiKey'] || !this.config['firebase']['projectId'] || !this.config['firebase']['messagingSenderId'] || !this.config['firebase']['appId']) {
        console.error('Missing Firebase keys');
        return false;
      }
      const config = this.config['firebase'] as Firebase;
      config.authDomain = config.authDomain || `${config.projectId}.firebaseapp.com`;
      config.storageBucket = config.storageBucket || `${config.projectId}.firebasestorage.app`;
      return config;
    }
    return false;
  }
}

type Analytics = { gtag?: string, clarity?: string; };
type ColorScheme = {
  background?: 'wood';
  main?: string;
  secondary?: string;
  tertiary?: string;
  quaternary?: string;
  light?: {
    main?: string;
    secondary?: string;
    tertiary?: string;
    quaternary?: string;
  };
};
type Socials = { facebook?: string, instagram?: string, threads?: string; };
type Firebase = { apiKey?: string, authDomain?: string, projectId?: string, storageBucket?: string, messagingSenderId?: string, appId?: string; };
