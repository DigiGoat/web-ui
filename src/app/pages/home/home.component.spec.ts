import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ConfigService } from '../../services/config/config.service';
import { ConfigServiceMock, EmptyConfigServiceMock } from '../../services/config/config.service.mock';
import { HomeComponent } from './home.component';


jest.mock('../../services/config/config.service');
const configService = jest.mocked(ConfigService);
describe('HomeComponent', () => {
  let component: HomeComponent;
  let fixture: ComponentFixture<HomeComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    configService.mockClear();
    await TestBed.configureTestingModule({
      declarations: [HomeComponent],
      providers: [ConfigService],
    })
      .compileComponents();
    fixture = TestBed.createComponent(HomeComponent);
    TestBed.inject(ConfigService);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
  });
  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });
  describe('With a config', () => {
    beforeEach(() => {
      configService.mockImplementation(() => {
        return ConfigServiceMock;
      });
      fixture.detectChanges();
    });
    it('should call ConfigService once', () => {
      expect(configService).toHaveBeenCalledTimes(1);
    });
    it('should have a title', () => {
      const element = html.querySelector('[test-id=title]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe(ConfigServiceMock.homeTitle);
    });
    it('should have an owner', () => {
      const element = html.querySelector('[test-id=owner]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe(ConfigServiceMock.owner);
    });
    it('should have an email', () => {
      const element = html.querySelector('[test-id=email]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe(`<a href="mailto:${ConfigServiceMock.email}">${ConfigServiceMock.email}</a>`);
    });
    it('should have a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe(ConfigServiceMock.homeDescription);
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('Without a config', () => {
    beforeEach(() => {
      configService.mockImplementation(() => {
        return EmptyConfigServiceMock;
      });
      fixture.detectChanges();
    });
    it('should call ConfigService once', () => {
      expect(configService).toHaveBeenCalledTimes(1);
    });
    it('should not have a title', () => {
      const element = html.querySelector('[test-id=title]');
      expect(element?.innerHTML).toBeFalsy();
    });
    it('should not have an owner', () => {
      const element = html.querySelector('[test-id=owner]');
      expect(element?.innerHTML).toBeFalsy();
    });
    it('should not have an email', () => {
      const element = html.querySelector('[test-id=email]');
      expect(element?.innerHTML).toBe('<a href="mailto:"></a>');
    });
    it('should not have a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('');
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
});
