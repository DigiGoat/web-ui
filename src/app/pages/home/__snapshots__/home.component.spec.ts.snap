// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeComponent With a config should match the snapshot 1`] = `
<app-home
  config={[Function Object]}
  description={[Function String]}
  meta={[Function _Meta]}
>
  <div
    class="w-100 text-center"
  >
    <img
      alt="Farm Logo"
      class="text-center"
      height="300"
      ngsrc="./assets/icons/android-chrome-192x192.png"
      priority=""
      width="300"
    />
  </div><h1
    class="text-center"
    test-id="title"
  >
    TEST_TITLE
  </h1><h5
    class="text-center"
    test-id="owner"
  >
    TEST_OWNER
  </h5><h6
    class="text-center"
    test-id="email"
  >
    <a
      href="mailto:TEST_EMAIL"
    >
      TEST_EMAIL
    </a>
  </h6><div
    class="container"
  >
    <p
      class="text-center"
      test-id="description"
    >
      TEST_DESCRIPTION
    </p>
  </div>
</app-home>
`;

exports[`HomeComponent Without a config should match the snapshot 1`] = `
<app-home
  config={[Function Object]}
  description=""
  meta={[Function _Meta]}
>
  <div
    class="w-100 text-center"
  >
    <img
      alt="Farm Logo"
      class="text-center"
      height="300"
      ngsrc="./assets/icons/android-chrome-192x192.png"
      priority=""
      width="300"
    />
  </div><h1
    class="text-center"
    test-id="title"
  >
  </h1><h5
    class="text-center"
    test-id="owner"
  >
  </h5><h6
    class="text-center"
    test-id="email"
  >
    <a
      href="mailto:"
    >
    </a>
  </h6><div
    class="container"
  >
    <p
      class="text-center"
      test-id="description"
    />
  </div>
</app-home>
`;
